### PR TITLE
DD4hep: DD World Built from a DD4hep-based Geometry

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

* DDWorld to access G4 geometry
* Converter to G4 geometry test

#### PR validation:

@mrodozov - this PR will not compile until the dd4hep-geant4 tool is setup correctly. When it does, it can be merged.

#### if this PR is a backport please specify the original PR:

no back port is needed

### CODE REVIEW

**Comment by mrodozov** [comment]
@ianna I fixed the include but now the code complains about "DDG4/Detector.h" not found in the DDG4 includes: 
https://github.com/cms-sw/cmssw/pull/26383/files#diff-19bce7dd0d1ec0083d7be74adfb4e755R6
is it DD4hep/Detector.h ?

**Comment by ianna** [comment]
Thanks @mrodozov ! Please, let me know when can I test it.

**Comment by mrodozov** [comment]
please test

**Comment by mrodozov** [comment]
I've setup latest stuff from dd4hep and now it doesn't break as before but it's missing a file:
DetectorDescription/DDCMS/interface/DetectorDescriptionRcd.h No such file or directory
To have in mind. 

DD4hep update PR is here: https://github.com/cms-sw/cmsdist/pull/4859

**Comment by ianna** [comment]
@mrodozov - yes, the PR needs a re-base

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
assign simulation
assign geometry

**Comment by ianna** [comment]
@fabiocos - this is a new package in Simulation with dependency on a dd4hep DDG4 static library.

**Comment by civanch** [comment]
@ianna , I am a bit confused. If it is a test, why these are producers in plugin directory instead of test directory? What is the purpose of addition of watchers? What they intend to watch?

File: SimG4Core/DD4hepGeometry/interface/DD4hep_DDDWorld.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  class DDDetector;
+  
+  class DDDWorld {
+
+  public:
+    DDDWorld(const cms::DDDetector*);
+    ~DDDWorld();
+    static void SetAsWorld(G4VPhysicalVolume* pv);
+    static void WorkerSetAsWorld(G4VPhysicalVolume* pv);
```
Comment (civanch) [inline SimG4Core/DD4hepGeometry/interface/DD4hep_DDDWorld.h]: @ianna , it is the same world, why making different methods WorkerSetAsWorld(G4VPhysicalVolume* pv) and GetWorldVolumeForWorker() ? I would remove both.

File: SimG4Core/DD4hepGeometry/plugins/DD4hep_DDTestG4Geometry.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  g4Geo.debugElements = true;
+  g4Geo.debugShapes = true;
+  g4Geo.debugVolumes = true;
+  g4Geo.debugPlacements = true;
+  g4Geo.create(detector.world());
+  
+  LogVerbatim("Geometry") << "Done.";
+}
+
+DEFINE_FWK_MODULE(DDTestG4Geometry);
```
Comment (civanch) [inline SimG4Core/DD4hepGeometry/plugins/DD4hep_DDTestG4Geometry.cc]: @ianna , I would move this to test from plugin.

File: SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+class DDDWorld;
+class G4RunManagerKernel;
+class SimTrackManager;
+
+using namespace edm;
+using namespace std;
+using namespace cms;
+
+static
+void createWatchers(const ParameterSet& iP, SimActivityRegistry& iReg,
```
Comment (civanch) [inline SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc]: @ianna , why we need this?

File: SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+      m_sensCaloDets.swap(sensDets.second);
+    }
+
+    LogInfo("G4GeometryESProducer") << " Sensitive Detector building finished; found " 
+				<< m_sensTkDets.size()
+				<< " Tk type Producers, and " << m_sensCaloDets.size() 
+				<< " Calo type producers ";
+  }
+
+  for(Producers::iterator itProd = m_producers.begin();itProd != m_producers.end();
```
Comment (civanch) [inline SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc]: @ianna , what is expected from this loop?

**Comment by civanch** [comment]
@ianna , if this package depends on geant4core, than geant4core cannot depend on DD4Hep. We have circular dependency.  In the current situation, SimG4Core depends on DetectorDescription/Core. I would expect we will have the same thing. For today we have SimG4Core/Geometry/test which is intended for testing of geometry only. Why not reuse the same sub-package or creating a new one?

**Comment by ianna** [comment]
This package is intended to test sim dependency on dd4hep. The DDDWorld is a copy of our DDDWorld with an exception that it is in a cms namespace. So is the other class. I was wondering myself about implementations and took liberty to drop some assignments which looked useless to me. 

If there’s a circular dependency, we have to address it.

My question is if the package affects the sim jobs in any way.

File: SimG4Core/DD4hepGeometry/interface/DD4hep_DDDWorld.h (hunk 1 above)
Comment (ianna) [reply SimG4Core/DD4hepGeometry/interface/DD4hep_DDDWorld.h]: Good question, I’d gladly remove the code

File: SimG4Core/DD4hepGeometry/plugins/DD4hep_DDTestG4Geometry.cc (hunk 2 above)
Comment (ianna) [reply SimG4Core/DD4hepGeometry/plugins/DD4hep_DDTestG4Geometry.cc]: Ok

File: SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc (hunk 3 above)
Comment (ianna) [reply SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc]: Good question

File: SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc (hunk 4 above)
Comment (ianna) [reply SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc]: Yet another good question:-)

**Comment by civanch** [comment]
@ianna , current OscarMTProducer (which do real simulation) access would volume via old DD scheme of CMSSW. Your new code seems to provide an access to the world volume. For the test it is fine. It will be good to remove things, which are not connected with geometry in this new producer, this will ensure that we have no circular dependencies (after second look, I am not sure if we really have).

File: SimG4Core/DD4hepGeometry/interface/DD4hep_DDDWorld.h (hunk 1 above)
Comment (ianna) [reply SimG4Core/DD4hepGeometry/interface/DD4hep_DDDWorld.h]: @civanch - all plugins ought to be in a plugin directory unrelated of their purpose. The Sim code is been copied as is to make sure that comparison is fair. I'll do a thorough cleanup at the later stage.

**Comment by ianna** [comment]
@civanch - there is no circular dependency. This package is needed to move further, e.g. to make a new OscarMTProducerFromDD4hep. However, it would help if it's integrated in an IB.

**Comment by fabiocos** [comment]
@ianna @civanch this PR looks now entirely embedded into #26740, which is fully approved, so I would say it may be closed

**Comment by ianna** [comment]
@fabiocos - yes, it's in #26740.

**Pull request closed without merging**